src/olympia/versions/tests/test_models.py

```diff
@@ -920,6 +920,45 @@ def test_pending_rejection_property(self):
         flags.update(pending_rejection=in_the_past)
         assert version.pending_rejection == in_the_past
 
+    def test_pending_rejection_by_property(self):
+        addon = Addon.objects.get(id=3615)
+        version = addon.current_version
+        user = user_factory()
+        # No flags: None
+        assert version.pending_rejection_by is None
+        # Flag present, value is None (default): None.
+        flags = VersionReviewerFlags.objects.create(version=version)
+        assert flags.pending_rejection_by is None
+        assert version.pending_rejection_by is None
+        # Flag present, value is a user.
+        flags.update(pending_rejection_by=user)
+        assert version.pending_rejection_by == user
+
+    def test_pending_rejection_by_cleared_when_pending_rejection_cleared(self):
+        addon = Addon.objects.get(id=3615)
+        version = addon.current_version
+        user = user_factory()
+        flags = VersionReviewerFlags.objects.create(
+            version=version,
+            pending_rejection=self.days_ago(1),
+            pending_rejection_by=user,
+        )
+        assert flags.pending_rejection
+        assert flags.pending_rejection_by == user
+        assert not flags.needs_human_review_by_mad
+
+        # Update, but do not clear pending_rejection. Both should remain.
+        flags.update(needs_human_review_by_mad=True)
+        flags.save()
```

Comment:
mmm, but what if there is _only_ a `flags.update(...)`?  You don't need to `.save()` after an update - it writes to the database.

Reply:
I'm glad you asked about this, because I don't quite understand it myself. I originally wrote the test with just `update` and no `save` afterwards, but the test was failing. It didn't seem that `update` was triggering the custom `save` method. This led me to believe that `update` doesn't save the record, but if it does then maybe there's a problem with what I've done here. I assumed that `save` was always called (for either a create or an update), but maybe it isn't? Or maybe it's related to this running in a test? Can you advise?

Reply:
`update` ends up calling the queryset's `update` method, which doesn't call save on the individual instances.  You could override `update` on the `VersionFlags` model too, but that wouldn't address using any of the queryset update methods directly - which I see is typically how it's used.

Adding stuff to `save` is (unfortunately) not a complete solution - that'd be database constraints.  There is some limited support for constraints now in mysql 8.0.16, and apparently in django3.2 they're supported, so you could try adding a `CheckConstraint` to the model? https://docs.djangoproject.com/en/3.2/ref/models/constraints/#django.db.models.CheckConstraint

(And change everywhere in the code where we currently clear `pending_rejection`; clearing `pending_rejection_user` too.)

Reply:
Ok, so my assumption that overriding `save` would "just work" and get fired any time an instance is changed and saved to the database is incorrect. I guess that means there is no point in having that code in `save`, and, as you say, we need to make sure we manually clear `pending_rejection_by` anytime we are also clearing `pending_rejection`. The database constraint would protect us from writing invalid data to the database, but we'd still need to manually clear both fields to avoid falling afoul of that constraint, so I'm not sure how much value it really adds.

@eviljeff  Is what I have written above accurate?

Reply:
To ensure your custom behavior would apply, you can put it in a [`pre_save`](https://docs.djangoproject.com/en/3.2/ref/signals/#pre-save) callback. `save()` triggers it, and our custom `update()` method does too, unless you really don't want it to (you can pass `_signal=False` to prevent that).

Reply:
I forgot about pre_save - yeah, that'd be an alternative to changing all the places we currently set `pending_rejection`.  A constraint would still be ideal though.

Reply:
Yeah `pre_save` would make the code work, the constraint would ensure the database is always in a clean state. Both could be useful here.

Reply:
Thanks @diox. I changed it to use `pre_save` and it works well. I can remove the `save` from my test and just doing an `update` triggers it. I will look into adding the constraint as well.

Do you both feel that the test I have for `test_pending_rejection_by_cleared_when_pending_rejection_cleared` is sufficient for this new `pre_save` code, or should I add assertions to other tests where `pending_rejection` is cleared to assert that `pending_rejection_by` is also cleared?